### PR TITLE
Enable/disable dynamic/static shared memory

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by lorc** [comment]
I'm okay with that. `CFG_DYN_SHM_CAP` was an easy way to test Linux kernel driver with and without dynamic SHM. New approach is much better.

File: core/arch/arm/kernel/generic_boot.c [hunk 1]
```diff
@@ -1060,7 +1062,9 @@ static void init_primary_helper(unsigned long pageable_part,
 	thread_init_per_cpu();
 	init_sec_mon(nsec_entry);
 	init_external_dt(fdt);
+#ifdef CFG_CORE_DYN_SHM
 	discover_nsec_memory();
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_boot.c]: I'll fix

File: core/arch/arm/mm/mobj_dyn_shm.c [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+static unsigned int reg_shm_map_lock = SPINLOCK_UNLOCK;
+
+static struct mobj_reg_shm *to_mobj_reg_shm(struct mobj *mobj);
+
+static TEE_Result mobj_reg_shm_get_pa(struct mobj *mobj, size_t offst,
+				      size_t granule, paddr_t *pa)
+{
+	struct mobj_reg_shm *mobj_reg_shm = to_mobj_reg_shm(mobj);
+	size_t full_offset;
+	paddr_t p;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_dyn_shm.c]: I'll fix all functions in this file.

File: core/arch/arm/mm/mobj_dyn_shm.c [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+	case 0:
+		p = mobj_reg_shm->pages[full_offset / SMALL_PAGE_SIZE] +
+			(full_offset & SMALL_PAGE_MASK);
+		break;
+	case SMALL_PAGE_SIZE:
+		p = mobj_reg_shm->pages[full_offset / SMALL_PAGE_SIZE];
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_dyn_shm.c]: OK

File: core/arch/arm/mm/mobj_dyn_shm.c [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+	s = mobj_reg_shm_size(num_pages);
+	if (!s)
+		return NULL;
+	mobj_reg_shm = calloc(1, s);
+	if (!mobj_reg_shm)
+		return NULL;
+
+	mobj_reg_shm->mobj.ops = &mobj_reg_shm_ops;
+	mobj_reg_shm->mobj.size =  num_pages * SMALL_PAGE_SIZE;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_dyn_shm.c]: OK

File: core/arch/arm/mm/mobj_dyn_shm.c [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+	mobj_reg_shm->mobj.phys_granule = SMALL_PAGE_SIZE;
+	mobj_reg_shm->cookie = cookie;
+	mobj_reg_shm->guarded = true;
+	mobj_reg_shm->num_pages = num_pages;
+	mobj_reg_shm->page_offset = page_offset;
+	memcpy(mobj_reg_shm->pages, pages, sizeof(*pages) * num_pages);
+	refcount_set(&mobj_reg_shm->refcount, 1);
+
+	/* Insure loaded references match format and security constraints */
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_dyn_shm.c]: OK

File: core/arch/arm/mm/mobj_dyn_shm.c [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+	for (i = 0; i < num_pages; i++) {
+		if (mobj_reg_shm->pages[i] & SMALL_PAGE_MASK)
+			goto err;
+
+		/* Only Non-secure memory can be mapped there */
+		if (!core_pbuf_is(CORE_MEM_NON_SEC, mobj_reg_shm->pages[i],
+					SMALL_PAGE_SIZE))
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj_dyn_shm.c]: OK

**Comment by jenswi-linaro** [comment]
Update

File: core/arch/arm/kernel/generic_boot.c [hunk 7]
```diff
@@ -818,6 +796,29 @@ static int add_res_mem_dt_node(struct dt_descriptor *dt, const char *name,
 	return 0;
 }
 
+#ifdef CFG_CORE_DYN_SHM
+static uint64_t get_dt_val_and_advance(const void *data, size_t *offs,
+				       uint32_t cell_size)
+{
+	uint64_t rv;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: can you init by the way. here and below.

**Comment by jenswi-linaro** [comment]
Comment addressed, tags applied

**Comment by jenswi-linaro** [comment]
The checkpatch warning is a false positive.

**Comment by jenswi-linaro** [comment]
Documentation updated in https://github.com/OP-TEE/optee_docs/pull/20

**Comment by jenswi-linaro** [comment]
Ping?

**Pull request merged**